providers/implementations/signature/ecdsa_sig.c

```diff
@@ -682,3 +856,224 @@ const OSSL_DISPATCH ossl_ecdsa_signature_functions[] = {
+static int ecdsa_sigalg_signverify_init(void *vctx, void *vec,
+                                      OSSL_FUNC_signature_set_ctx_params_fn *set_ctx_params,
+                                      const OSSL_PARAM params[],
+                                      const char *mdname,
+                                      int operation, const char *desc)
+{
+    PROV_ECDSA_CTX *ctx = (PROV_ECDSA_CTX *)vctx;
+
+    if (!ossl_prov_is_running())
+        return 0;
+
+    if (!ecdsa_signverify_init(vctx, vec, set_ctx_params, params, operation,
+                               desc))
+        return 0;
+
+    if (!ecdsa_setup_md(ctx, mdname, NULL, desc))
+        return 0;
+
+    ctx->flag_sigalg = 1;
+    ctx->flag_allow_md = 0;
+
+    if (ctx->mdctx == NULL) {
+        ctx->mdctx = EVP_MD_CTX_new();
+        if (ctx->mdctx == NULL)
+            goto error;
+    }
+
+    if (!EVP_DigestInit_ex2(ctx->mdctx, ctx->md, params))
+        goto error;
+
+    return 1;
+
+ error:
+    EVP_MD_CTX_free(ctx->mdctx);
+    ctx->mdctx = NULL;
+    return 0;
+}
+
+static const char **ecdsa_sigalg_query_key_types(void)
+{
+    static const char *keytypes[] = { "EC", NULL };
+
+    return keytypes;
+}
+
+static const OSSL_PARAM settable_sigalg_ctx_params[] = {
+    OSSL_PARAM_octet_string(OSSL_SIGNATURE_PARAM_SIGNATURE, NULL, 0),
+    OSSL_PARAM_uint(OSSL_SIGNATURE_PARAM_NONCE_TYPE, NULL),
+    OSSL_FIPS_IND_SETTABLE_CTX_PARAM(OSSL_SIGNATURE_PARAM_FIPS_KEY_CHECK)
+    OSSL_FIPS_IND_SETTABLE_CTX_PARAM(OSSL_SIGNATURE_PARAM_FIPS_DIGEST_CHECK)
+    OSSL_PARAM_END
+};
+
+static const OSSL_PARAM *ecdsa_sigalg_settable_ctx_params(void *vctx,
+                                                        ossl_unused void *provctx)
+{
+    PROV_ECDSA_CTX *ctx = (PROV_ECDSA_CTX *)vctx;
+
+    if (ctx != NULL && ctx->operation == EVP_PKEY_OP_VERIFYMSG)
+        return settable_sigalg_ctx_params;
+    return NULL;
+}
+
+static int ecdsa_sigalg_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    PROV_ECDSA_CTX *ctx = (PROV_ECDSA_CTX *)vctx;
+    const OSSL_PARAM *p;
+
+    if (ctx == NULL)
+        return 0;
+    if (params == NULL)
+        return 1;
+
+    if (!OSSL_FIPS_IND_SET_CTX_PARAM(ctx, OSSL_FIPS_IND_SETTABLE0, params,
+                                     OSSL_SIGNATURE_PARAM_FIPS_KEY_CHECK))
+        return  0;
+    if (!OSSL_FIPS_IND_SET_CTX_PARAM(ctx, OSSL_FIPS_IND_SETTABLE1, params,
+                                     OSSL_SIGNATURE_PARAM_FIPS_DIGEST_CHECK))
+        return  0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_NONCE_TYPE);
+    if (p != NULL
+        && !OSSL_PARAM_get_uint(p, &ctx->nonce_type))
+        return 0;
+
+    if (ctx->operation == EVP_PKEY_OP_VERIFYMSG) {
+        p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_SIGNATURE);
+        if (p != NULL) {
+            OPENSSL_free(ctx->sig);
+            ctx->sig = NULL;
+            ctx->siglen = 0;
+            if (!OSSL_PARAM_get_octet_string(p, (void **)&ctx->sig,
```

Comment:
potentially params->data could be NULL here..

Reply:
Would you say that's a permissible scenario?  Is it wrong to return an error indication in that case?

Reply:
It was just an observation now that we are passing signatures around, which we werent before - and for other new algorithms at least it could be NULL in the final. So maybe it should even get here to start with

Reply:
"for new algorithms...", but that's not what this is, is it?  I don't see how a NULL signature is valid for ECDSA.